src/main/java/org/wso2/integration/inbound/PulsarMessageConsumer.java

```diff
@@ -0,0 +1,656 @@
+
+    private String topicNames;
+    private String topicsPattern;
+    private String subscriptionName;
+    private RegexSubscriptionMode subscriptionTopicsMode;
+    private SubscriptionType subscriptionType;
+    private SubscriptionInitialPosition subscriptionInitialPosition;
+    private String consumerName;
+
+
+    private Integer messageWaitTimeout = 1;
+    private Long ackTimeoutMillis;
+    private Long nackRedeliveryDelayMillis;
+
+    private Integer priorityLevel;
+    private Integer receiverQueueSize;
+    private Integer maxTotalReceiverQueueSizeAcrossPartitions;
+
+    private String dlqTopic;
+    private Integer dlqMaxRedeliverCount;
+
+    // Configuration for chunked messages
+    private Boolean autoAckOldestChunkedMessageOnQueueFull;
+    private Integer maxPendingChunkedMessage;
+    private Long expiryTimeOfIncompleteChunkedMessageMillis;
+
+    private Boolean autoUpdatePartitions;
+    private Integer autoUpdatePartitionsIntervalSeconds;
+    private Boolean replicateSubscriptionState;
+    private Boolean readCompacted;
+    private SortedMap<String, String> properties;
+
+    private Boolean receiveSync = true;
+
+    // Batching related configurations
+    private Boolean batchReceiveEnabled;
+    private Integer batchingMaxMessages;
+    private Integer batchingMaxBytes;
+    private Integer batchingTimeout;
+    private Boolean batchIndexAcknowledgmentEnabled;
+    private String contentType;
+
+    // Thread pool for processing messages asynchronously
+    int corePoolSize = 4;       // minimum number of threads
+    int maxPoolSize = 8;        // maximum threads
+    int queueCapacity = 100;    // max number of tasks in the queue
+
+    ExecutorService boundedExecutor = new ThreadPoolExecutor(
+            corePoolSize,
+            maxPoolSize,
+            60L, TimeUnit.SECONDS,
+            new ArrayBlockingQueue<>(queueCapacity), // bounded queue
+            Executors.defaultThreadFactory(),
+            // Policy: what happens when queue is full and max threads are busy
+            new ThreadPoolExecutor.CallerRunsPolicy() // Runs task in calling thread
+    );
+
+    public PulsarMessageConsumer(Properties properties, String name, SynapseEnvironment synapseEnvironment,
+                                 long scanInterval, String injectingSeq, String onErrorSeq, boolean coordination,
+                                 boolean sequential) {
+
+        super(properties, name, synapseEnvironment, scanInterval, injectingSeq, onErrorSeq, coordination, true);
+        configuration = PulsarUtils.getConnectionConfigFromProperties(properties);
+        getConsumerConfigFromProperties(properties);
+    }
+
+    @Override
+    public Object poll() {
+
+        try {
+            if (consumer == null) {
+                consumer = createConsumer(configuration);
+            }
+            consumeMessages();
+
+        } catch (PulsarClientException e) {
+            throw new RuntimeException(e);
```

Comment:
Is it safe to throw runtime exception here?

Reply:
Changed here to log an error message